chainer/links/loss/hierarchical_softmax.py

```diff
@@ -349,6 +396,85 @@ def create_huffman_tree(word_counts):
 
         return q.get()[2]
 
+    def sample(self, x):
+        """Sample an example for a given input form the tree.
+
+        Args:
+            x (~chainer.Variable): Input variable for sample word ids.
+
+        Returns:
+            List of word indexes.
+
+        .. admonition:: Example
+
+            Let an input vector ``x`` be:
+
+            >>> x = np.array([[0.2, 0.2, 0.3], [0.1, 0.3, 0.1]], dtype='f')
+            >>> hsm.sample(Variable(x))
+            [0, 3]
+
+
+        """
+        if len(self.tree) == 0:
+            raise ValueError('Empty tree')
+
+        xp = cuda.get_array_module(*x)
+        parent2child = xp.array(self._func.parent2child, dtype=xp.int32)
+        batchsize = x.shape[0]
+        start_ids = xp.zeros(batchsize, 'i')
+        list_next_ids = []
+        list_choose_ids = []
+
+        def _sigmoid(x):
+            half = x.dtype.type(0.5)
+            return self.xp.tanh(x * half) * half + half
+
+        while True:
+            w = self.W.data[start_ids]
+            x_t = xp.transpose(x.data)
+            score = xp.sum(xp.dot(w, x_t), axis=1)
+            prob_left = _sigmoid(score)[:, None]
+            prob_right = xp.ones_like(prob_left) - prob_left
+            prob = xp.concatenate([prob_left, prob_right], axis=1)
+
+            choosed_idx = xp.argmax(xp.random.gumbel(size=prob.shape) + prob,
+                                    axis=1)
+            rows = xp.arange(batchsize, dtype=xp.int32)
+            columns = choosed_idx
+
+            list_next_ids.append(start_ids)
+            list_choose_ids.append(choosed_idx)
+
+            next_ids = parent2child[start_ids][rows, columns]
+            next_ids = xp.where(next_ids != LEAF, next_ids, FINISH_SAMPLING)
+
+            # check whether all nodes are LEAF.
+            if xp.all(next_ids == FINISH_SAMPLING):
+                # if all nodes will reach leaf, then finish sampling.
+                break
+            start_ids = next_ids
+
+        def xp_stack_func(x):
+            return xp.reshape(xp.concatenate(x), (-1, batchsize)).T
+
+        next_ids = xp_stack_func(list_next_ids)
+        choose_ids = xp_stack_func(list_choose_ids)
+        lengths = xp.argmax(next_ids == FINISH_SAMPLING, axis=1)
+        max_length = choose_ids.shape[1]
+        lengths = xp.where(lengths == 0, max_length, lengths)
+        sampling = []
+        for length, path in zip(lengths, choose_ids):
+            sampling.append(path[0:length])
+
+        # Find word id from tree.
+        output = []
+        for sampling_lst in sampling:
```

Comment:
Is it possible to find leaf id in mini-batch using `choose_ids` and `parent2child`?